protocol_grpc.go

```diff
@@ -539,27 +539,7 @@ func (hc *grpcHandlerConn) Close(err error) (retErr error) {
 	// as HTTP trailers. (If we had frame-level control of the HTTP/2 layer, we
 	// could send trailers-only responses as a single HEADER frame and no DATA
 	// frames, but net/http doesn't expose APIs that low-level.)
-	if !hc.wroteToBody {
-		// This block works around a bug in x/net/http2. Until Go 1.20, trailers
-		// written using http.TrailerPrefix were only sent if either (1) there's
-		// data in the body, or (2) the innermost http.ResponseWriter is flushed.
-		// To ensure that we always send a valid gRPC response, even if the user
-		// has wrapped the response writer in net/http middleware that doesn't
-		// implement http.Flusher, we must pre-declare our HTTP trailers. We can
-		// remove this when Go 1.21 ships and we drop support for Go 1.19.
-		for key := range mergedTrailers {
-			addHeaderCanonical(hc.responseWriter.Header(), headerTrailer, key)
-		}
```

Comment:
This is the only part that is safe to remove. You still need the rest, to actually set the trailer values, but will need to use `http.TrailerPrefix` in the key to make it clear they are trailers (since they weren't already advertised in a `Trailer` header).

Reply:
I think it's safe to remove all. Trailers are written with the prefix below.

Reply:
Ah, my bad. I somehow overlooked the `return nil` on line 561